rfd/0181-pg-database-access-webui.md

```diff
@@ -0,0 +1,322 @@
+}
+```
+
+These new fields are used to:
+- Change the "Connect" button behavior to present a connect modal when web UI
+  is supported.
+- Present the available database roles when applicable.
+
+In addition, a new page will be added to handle the database session
+console/terminal (`/web/cluster/:clusterId/console/db/:sid`). This page will be
+very similar to the SSH session console, except for the file transfer/upload
+icons bar, which will be removed.
+
+#### Proxy
+
+A new web handler will be added to handle the interactive session socket
+connection: `/webapi/sites/:site/databases/term`. We cannot provide the session
+parameters since web socket connections are started as GET requests. Instead,
+when the web socket connection is established, the front end will forward a
+message containing the session parameters. This is the same flow used by
+Kubernetes exec sessions.
+
+The following struct represents the session request:
+
+```go
+// DatabaseSessionRequest describes a request to create a web-based terminal
+// database session.
+type DatabaseSessionRequest struct {
+  // DatabaseResourceID is the database resource ID the user will be connected.
+  DatabaseResourceID string `json:"db_resource_id"`
+  // DatabaseName is the database name the session will use.
+  DatabaseName string `json:"db_name"`
+  // DatabaseUser is the database user used on the session.
+  DatabaseUser string `json:"db_user"`
+  // DatabaseRoles are ratabase roles that will be attached to the user when connecting to the database.
+  DatabaseRoles []string `json:"db_roles"`
+}
+```
+
+After the connection is established and the first message is received, the proxy
+will generate a user certificate containing `RouteToDatabase` using one of
+`PerformMFACeremony` or `GenerateUserCerts`.
+
+Note: The MFA flows will follow the existing functions (used by Kubernetes
+execs, desktops, and SSH sessions).
+
+Once the certificates are generated, the proxy will connect to the target
+database server. This flow resembles what the database proxy does: Choose a
+database server among a list of available servers to forward the connection.
+After the connection is established, a new instances of the Teleport's
+PostgreSQL REPL is created, and the session starts.
+
+#### PostgreSQL REPL
+
+The proxy will include a PostgreSQL Read-Eval-Print Loop (REPL) that interacts
+with the databases served by database servers. The REPL will read the user input
+per line and decide whether or not to execute the command (as the REPL will
+support multi-line commands). The rule follows:
+- If the line starts with a backslash command, execute it.
+- Send the line to the target database if it terminates in a semicolon (`;`).
+
+If the line read doesn't meet the above criteria, it is added to an internal
+buffer and is accumulated until it is ready to be executed.
+
+The REPL will interact with the WebSocket connection using the existent
+`terminal` (`lib/web/terminal/terminal.go`) structs.
```

Comment:
I've tried a few routes on my PoCs. Here are my thoughts on this (I'll later add this discussion to the RFD for future reference):

> My current concerns is that and sanding and reading raw bytes from the frontend might limit the functionality for extensibility of Database Web UI Client.

> For example, crafting a custom PSQL client in a WASM module shouldn’t be difficult. We could reuse the [usql](https://github.com/xo/usql/tree/master) library and add more complex logic, such as query completion, to gain better control over web UI CLI behavior. Additionally, we could implement custom protocols for AssistAI, particularly when the database schema is known.

I don't think we would be limited in bringing new features to the product. The main difference is how we approach them: solving entirely on the Web UI, doing most of the work on the proxy, and only adding minor web UI changes.

From my testing with something similar, the increase in Web UI complexity didn't pay off. This was mainly because we could do the same flows on the proxy and quickly bring those flows into `tsh` (for example, to remove the `tsh` dependency on `psql` to connect to PostgreSQL databases). It is worth mentioning that bringing the REPL to the client would still require a new connection on the proxy due to certificate generation and the web browser's inability to establish a connection other than HTTP directly. We would need at least a PostgreSQL-over-WebSocket with user certificate generation.

> How does current approach will allow us to extend the WebUI database client and add support for more complex features

Generally speaking, we would require a deep look at how each feature fits into the product UX. For some features you mentioned (rich web UI and AI integration), this terminal-based interface might not be the best, and it would be better if their UX were more exercised before relying on the terminal-based flow. Still, it works great with more requests/reply flows with a different UI, such as a code editor (used [pglite REPL](https://pglite.dev/repl/) as a reference). In addition, the session management in those scenarios might also be different, and we can even decide in favor of a stateless (e.g., offering new endpoints to perform database queries). My takeaway is that we should take those additions a separate look and decide how to include them in our product and build a great UX.

As for the initial proposed UX (terminal-based), building the REPL on the proxy side gives us some advantages if/when we introduce features already present on our other terminal-based UX, such as moderated sessions (with typing feedback) and session resumption. In addition, autocompletion can also have a simple implementation (similar to bash/shell autocompletion), which would be done on the REPL itself.

Regarding implementation and code reuse, the structure established by the text-based session should be easily reused in those other scenarios.

> Or at lest craft custom wire protocol between FE and BE proxy for WebSocket connection that will give us more control.

Our current solution for handling WebSocket supports [adding different message types](https://github.com/gravitational/teleport/blob/714c1742578e77b8d1e874c42d4d2fbb9482cb46/proto/teleport/lib/web/terminal/envelope.proto#L33). We can still rely on that and have a custom UI. We might consider using a different envelope message focused on database access capabilities.

Reply:
I haven't looked much into WASM. could we just do `tsh`?

Personally, I agree with Gabriel that we have to look at advanced features individually and they might need different flows.

The main advantage of text-based is to reuse our term manager for moderated sessions (observer, peer, etc.) where participants just tap into the same stdin/stdout streams.

I also believe auto-completion is text-based and so it should be doable with the current implementation.